spopt/locate/base.py

```diff
@@ -0,0 +1,406 @@
+        range_client: range
+            range of demand points quantity
+        var_name: str
+            formatted string
+            demand variable name
+
+        Returns
+        -------
+        None
+        """
+        cli_vars = [
+            pulp.LpVariable(
+                var_name.format(i=i), lowBound=0, upBound=1, cat=pulp.LpInteger
+            )
+            for i in range_client
+        ]
+
+        setattr(obj, "cli_vars", cli_vars)
+
+    @staticmethod
+    def add_client_assign_integer_variable(
+        obj: T_FacModel, range_client, range_facility, var_name
+    ) -> None:
+        """
+
+        Parameters
+        ----------
+        obj: T_FacModel
+            bounded type of LocateSolver class
+        range_client: range
+            range of demand points quantity
+        range_facility: range
+            range of facility points quantity
+        var_name: str
+            formatted string
+            client assigning variable name
+
+        Returns
+        -------
+        None
+        """
+        cli_assgn_vars = [
+            [
+                pulp.LpVariable(
+                    var_name.format(i=i, j=j), lowBound=0, upBound=1, cat=pulp.LpInteger
+                )
+                for j in range_facility
+            ]
+            for i in range_client
+        ]
+
+        setattr(obj, "cli_assgn_vars", cli_assgn_vars)
+
+    @staticmethod
+    def add_weight_continuous_variable(obj: T_FacModel) -> None:
+        """
+
+        Parameters
+        ----------
+        obj: T_FacModel
+            bounded type of LocateSolver class
+
+        Returns
+        -------
+        None
+        """
+        weight_var = pulp.LpVariable("W", lowBound=0, cat=pulp.LpContinuous)
+
+        setattr(obj, "weight_var", weight_var)
+
+    @staticmethod
+    def add_set_covering_constraint(
+        obj: T_FacModel,
+        model: pulp.LpProblem,
+        ni: np.array,
+        range_facility: range,
+        range_client: range,
+    ) -> None:
+        """
+        set covering constraint:
+        n1_1 * fac_var1 + n1_2 * fac_var1 + ... + nij * fac_varj >= 1
+
+        Parameters
+        ----------
+        obj: T_FacModel
+            bounded type of LocateSolver class
+        model: pulp.LpProblem
+            optimization model problem
+        ni: np.array
+            two-dimensional array that defines candidate sites between facility points within a distance to supply {i}
+            demand point
+        range_facility: range
+            range of facility points quantity
+        range_client: range
+            range of demand points quantity
+        Returns
+        -------
+        None
+
+        """
+        try:
+            fac_vars = getattr(obj, "fac_vars")
+            for i in range_client:
+                model += (
+                    pulp.lpSum([ni[i][j] * fac_vars[j] for j in range_facility]) >= 1
+                )
+        except AttributeError:
```

Comment:
Where we purposely raise exceptions, we should ideally test to make sure they raise as expected. This (in general) applies to all situations where errors could be raise. See an example [here](https://github.com/pysal/spopt/blob/42520cc53d10c14a61ac8ba79bdd45e080281775/spopt/tests/test_random_regions.py#L157).